gitlab/client.py

```diff
@@ -1199,12 +1201,26 @@ def _query(
         self._total: Optional[str] = result.headers.get("X-Total")
 
         try:
-            self._data: List[Dict[str, Any]] = result.json()
+            data: List[Dict[str, Any]] = result.json()
         except Exception as e:
             raise gitlab.exceptions.GitlabParsingError(
                 error_message="Failed to parse the server message"
             ) from e
 
+        self._data = []
+        for item in data:
```

Comment:
Are we manually looping over items and adding them one by one for performance reasons here? Or could we just do a `list(OrderedDict.fromkeys(data))` and emit a single warning with the difference from the initial data vs. filtered?

Reply:
We could remove the loop and make the code more concise.  As far as complexity goes, doing a `set` difference still has a loop it's just abstracted away and done in C instead of Python.  I imagine the performance difference (loop is max 100 items) is negligible compared to the network API call.

I will change the code though and update this PR.

Reply:
Actually, looks like we can't use `set()` or `OrderedDict` on these as here the objects are still just unhashable dicts.

I like option 2 as it's more concise.  Also perhaps CPython has optimized list comprehensions where option 1 is a generic for loop.

If we only keep a single page's previous objects we can choose between these two options.

If we decide to keep all objects, we should instead use a `set()` of `int`s.  So maybe let's decide that first.

## option 1

One loop; calculate dupes and filter within a single loop

```python
duplicates = []
self._data = []
for item in data:
    if item in self._prev_page_objects:
        duplicates.append(item)
        continue
    self._data.append(item)
```

## option 2

Two loops; one to calculate dupes, another to do the filtering.

```python
# loop once to calculate dupes
duplicates = [o for o in data if o in self._prev_page_objects]
# loop again to remove the dupes
self._data = [o for o in data if o not in dupes]
```

## single warning

In both cases we can defer the warning until all duplicates are detected and emit a single warning.

```python
if duplicates:
    utils.warn(
        message=(
            f"During pagination duplicate object with id(s) "
            f"{[item['id'] for item in duplicates]} returned from Gitlab and filtered"
        ),
        category=UserWarning,
    )
```